homeassistant/components/qwikswitch.py

```diff
@@ -76,127 +79,98 @@ def is_on(self):
-        """Turn the device off."""
-        self._qsusb.set_value(self._id, 0)
-
-    @asyncio.coroutine
-    def async_turn_off(self, **kwargs):  # pylint: disable=unused-argument
+    async def async_turn_off(self, **_):
         """Turn the device off."""
         self._qsusb.set_value(self._id, 0)
 
 
-class QSSwitch(QSToggleEntity, SwitchDevice):
-    """Switch based on a Qwikswitch relay module."""
-
-    pass
-
-
-class QSLight(QSToggleEntity, Light):
-    """Light based on a Qwikswitch relay/dimmer module."""
-
-    @property
-    def brightness(self):
-        """Return the brightness of this light (0-255)."""
-        return self._qsusb[self._id, 1] if self._dim else None
-
-    @property
-    def supported_features(self):
-        """Flag supported features."""
-        return SUPPORT_BRIGHTNESS if self._dim else None
-
-
-@asyncio.coroutine
-def async_setup(hass, config):
-    """Setup qwiskswitch component."""
-    from pyqwikswitch.async import QSUsb
+async def async_setup(hass, config):
+    """Qwiskswitch component setup."""
+    from pyqwikswitch.async_ import QSUsb
     from pyqwikswitch import (
-        CMD_BUTTONS, QS_CMD, QSDATA, QS_ID, QS_NAME, QS_TYPE, QSType)
+        CMD_BUTTONS, QS_CMD, QS_ID, QS_TYPE, QSType)
 
-    hass.data[DOMAIN] = {}
-
-    # Override which cmd's in /&listen packets will fire events
+    # Add cmd's to in /&listen packets will fire events
     # By default only buttons of type [TOGGLE,SCENE EXE,LEVEL]
-    cmd_buttons = config[DOMAIN].get(CONF_BUTTON_EVENTS, ','.join(CMD_BUTTONS))
-    cmd_buttons = cmd_buttons.split(',')
+    cmd_buttons = set(CMD_BUTTONS)
+    for btn in config[DOMAIN].get(CONF_BUTTON_EVENTS, []):
+        cmd_buttons.add(btn)
 
     url = config[DOMAIN][CONF_URL]
     dimmer_adjust = config[DOMAIN][CONF_DIMMER_ADJUST]
+    sensors = config[DOMAIN]['sensors']
+    switches = config[DOMAIN]['switches']
 
-    def callback_value_changed(qsdevices, key, new): \
-            # pylint: disable=unused-argument
-        """Update entiry values based on device change."""
-        entity = hass.data[DOMAIN].get(key)
-        if entity is not None:
-            entity.schedule_update_ha_state()  # Part of Entity/ToggleEntity
+    def callback_value_changed(_qsd, key, _val):
+        """Update entity values based on device change."""
+        hass.helpers.dispatcher.async_dispatcher_send(key)
 
     session = async_get_clientsession(hass)
     qsusb = QSUsb(url=url, dim_adj=dimmer_adjust, session=session,
                   callback_value_changed=callback_value_changed)
 
-    @callback
-    def async_stop(event):  # pylint: disable=unused-argument
-        """Stop the listener queue and clean up."""
-        nonlocal qsusb
-        qsusb.stop()
-        qsusb = None
-        hass.data[DOMAIN] = {}
-        _LOGGER.info("Waiting for long poll to QSUSB to time out")
+    # Discover all devices in QSUSB
+    if not await qsusb.update_from_devices():
+        raise PlatformNotReady
```

Comment:
This only has an effect if raised within `setup_platform`.

Reply:
Unfortunate, for now I will just remove and leave more clever recovery mechnism for a future PR